.github/workflows/test.yml

```diff
@@ -1,24 +1,113 @@
 name: "Arweave Tests"
 on:
+  workflow_dispatch:
   push:
     branches: ["**"]
 jobs:
-  test:
-    runs-on:
-      labels: ubuntu-latest-16-cores
+  common-test-suite:
+    runs-on: self-hosted
     steps:
-      - uses: actions/checkout@v2.3.4
+      - uses: actions/checkout@v4
         with:
           submodules: "recursive"
-      - uses: erlef/setup-beam@v1
-        with:
-          otp-version: "24.3"
-          rebar3-version: "3.20.0"
-      - name: Install linux dependencies
-        run: sudo apt-get install -y sqlite3 libgmp-dev
-      - name: Start empd
-        run: epmd -daemon
+      - name: Build arweave test sources
+        run: |
+          nix build .#packages.x86_64-linux.arweave
+          cp -Lr result _build
+          chmod -R u+w ./_build
+      - name: Start empd if needed
+        run: |
+          if ! pgrep -x "epmd" > /dev/null
+          then
+            epmd -relaxed_command_check -packet_timeout 30 -daemon
+          fi
       - name: CommonTest Suite
-        run: rebar3 ct --cover --sname=master@localhost
-      - name: Run Arweave Tests
-        run: ./bin/test
+        run: |
+          export PATH=$(pwd)/_build/erts/bin:$PATH
+          ct_run -dir ./_build/test/lib/arweave/test \
+            -noinput -no_auto_compile -logopts no_src \
+            -include $(pwd)/_build/test/lib/arweave/include \
+            -config $(pwd)/config/sys.config \
+            -pa $(echo $(pwd)/_build/test/rel/arweave/lib/*/ebin) \
+            -suite "ar_SUITE.erl" \
+            -sname master
+  eunit-tests:
+    runs-on: self-hosted
+    needs: common-test-suite
+    strategy:
+      fail-fast: true
+      max-parallel: 3
+      matrix:
+        core_test_mod: [
+            ar,
+            ar_block,
+            ar_block_cache,
+            ar_chunk_storage,
+            ar_data_sync_worker_master,
+            ar_deep_hash,
+            ar_diff_dag,
+            ar_ets_intervals,
+            ar_inflation,
+            ar_intervals,
+            ar_join,
+            ar_kv,
+            ar_merkle,
+            ar_mining_server,
+            ar_node,
+            ar_node_utils,
+            ar_nonce_limiter,
+            ar_packing_server,
+            ar_patricia_tree,
+            ar_peers,
+            ar_poa,
+            ar_pricing,
+            ar_retarget,
+            ar_serialize,
+            ar_storage,
+            ar_sync_buckets,
+            ar_tx,
+            ar_tx_db,
+            ar_unbalanced_merkle,
+            ar_util,
+            ar_weave,
+            # note we don't explicitly include *_tests.erl here
```

Comment:
I think we may need some of the *_tests.erl file - not all of them share the same name as an existing module

Reply:
and we have smoe that match an existing module, but that module itself doesn't have tests so isn't in the `core_test_mod` list

Reply:
actually it looks like that's a lot of *_tests.erl files...

Reply:
basically I think we need to include all the *_tests.erl file *except* for those that match the list above.